src/devices_models/devices/common/timeseries_constraint.jl

```diff
@@ -22,20 +22,43 @@ Constructs upper bound for given variable and time series data and a multiplier.
 """
 function device_timeseries_ub(psi_container::PSIContainer,
                               ts_data::Vector{Tuple{String, Int64, Float64, Vector{Float64}}},
+                              range_data::DeviceRange, #TODO: there is some repititve info in range_data and ts_data, and consistend order/size are assumed, we should combine into a single argument
                               cons_name::Symbol,
                               var_name::Symbol)
     time_steps = model_time_steps(psi_container)
     names = (v[1] for v in ts_data)
-    constraint = add_cons_container!(psi_container, cons_name, names, time_steps)
     variable = get_variable(psi_container, var_name)
+    ub_name = _middle_rename(cons_name, "_", "ub")
+    con_ub = add_cons_container!(psi_container, ub_name, names, time_steps)
+    add_lower_bound = !all(isempty.(range_data.additional_terms_lb))
+    if add_lower_bound
+        lb_name = _middle_rename(cons_name, "_", "lb")
+        con_lb = add_cons_container!(psi_container, lb_name, names, time_steps)
+    end
 
-    for t in time_steps, data in ts_data
-        name = data[1]
-        forecast = data[4][t]
-        multiplier = data[3]
-        constraint[name, t] = JuMP.@constraint(psi_container.JuMPmodel,
-                                    variable[name, t] <= multiplier*forecast)
-
+    for (ix, name) in enumerate(range_data.names)
+        data = ts_data[ix]
+        for t in time_steps
+            @assert name == data[1] # ensures that ts_data and range_data have same order
+            forecast = data[4][t]
+            multiplier = data[3]
+            expression_ub = JuMP.AffExpr(0.0, variable[name, t] => 1.0)
+            for val in range_data.additional_terms_ub[ix]
+                JuMP.add_to_expression!(expression_ub, 
+                                        get_variable(psi_container, val)[name, t])
+            end
+            con_ub[name, t] = JuMP.@constraint(psi_container.JuMPmodel,
+                                               expression_ub <= multiplier*forecast)
+            if add_lower_bound
+                expression_lb = JuMP.AffExpr(0.0, variable[name, t] => 1.0)
+                for val in range_data.additional_terms_lb[ix]
+                    JuMP.add_to_expression!(expression_lb, 
+                                            get_variable(psi_container, val)[name, t], -1.0)
+                end
+                con_lb[name, t] = JuMP.@constraint(psi_container.JuMPmodel,
+                                                expression_lb >= 0.0)
+            end
+        end
```

Comment:
There is a lot of code duplication across these functions which may cause future maintenance headaches.  Can some of the logic be consolidated?

Reply:
These functions need refactoring for #219. Definitely they can be simplified.